### DISCUSSION TITLE
Need help with secrets configuration

### DISCUSSION THREAD

**Discussion by amimas** [renovatebot/renovate#10197] [Request Help] [answered]
Hello, I'm running renovate using yarn.  I found from the renovate's own `package.json` that there's a config validator named `renovate-config-validator`. When I run it, I get an error message as shown below indicating that my config is invalid around "secrets" configuration.

```
 INFO: Validating config.js
ERROR: config.js contains errors
       "errors": [
         {
           "topic": "Configuration Error",
           "message": "Invalid configuration option: secrets.DOCKERHUB_USER_TOKEN"
         },
         {
           "topic": "Configuration Error",
           "message": "Invalid configuration option: secrets.QUAY_USER_TOKEN"
         }
       ]
```

Below is my `config.js`. I can't see what's wrong with the "secrets" configurations. 

```
{
  platform: 'gitlab',
  endpoint: 'https://***/api/v4/',
  token: '***',
  autodiscover: true,
  binarySource: 'docker',
  gitAuthor: '***',
  force: { automerge: false, enabled: true },
  logFile: 'renovate.log',
  logFileLevel: 'debug',
  onboarding: false,
  optimizeForDisabled: true,
  persistRepoData: false,
  prCommitsPerRunLimit: 0,
  printConfig: true,
  requireConfig: false,
  secrets: {
    DOCKERHUB_USER_TOKEN: '***',
    QUAY_USER_TOKEN: '***'
  },
  extends: [ 'config:base' ],
  assigneesFromCodeOwners: true,
  automerge: false,
  commitBodyTable: true,
  enabled: true,
  hostRules: [
    {
      matchHost: 'index.docker.io',
      token: '{{ secrets.DOCKERHUB_USER_TOKEN }}',
      authType: 'Basic'
    },
    {
      matchHost: 'quay.io',
      token: '{{ secrets.QUAY_USER_TOKEN }}',
      authType: 'Basic'
    }
  ],
  prFooter: 'This MR has been generated by [Renovate Bot](https://***/renovatebot/)',
  schedule: []
}
```

If I run renovate (using dry-run so far), it does run but I do get following errors around authenticate with docker hub. So, I'm not sure if my token is somehow incorrect or if there is really an error in the way I have things in `config.js`.

```
INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 50,
           "logContext": "PQuxo6OR4",
           "repository": "****",
           "branch": "renovate/go.mongodb.org-mongo-driver-1.x",
           "msg": "No docker.io/renovate/go releases found"
         },
         {
           "name": "renovate",
           "level": 50,
           "logContext": "PQuxo6OR4",
           "repository": "****",
           "branch": "renovate/github.com-caarlos0-env-6.x",
           "msg": "No docker.io/renovate/go releases found"
         }
       ]
```

**Comment by amimas** [comment]
I also get this error, which I can't reproduce consistently on my local machine. I've posted a small snippet of the full stack. Let me know if you need to see the full error. But notice here that the error message is indicating `authentication scheme: Bearer`, even though in my config posted above, the authentication type is set to `Basic`.


```json
WARN: Error obtaining docker token (repository=***)
       "registryHost": "https://index.docker.io",
       .....
         "response": {
           "statusCode": 400,
           "statusMessage": "Bad Request",
           "body": {"details": "unsupported authentication scheme: Bearer"},
           "headers": {
             "date": "Thu, 27 May 2021 21:25:27 GMT",
             "content-length": "56",
             "content-type": "text/plain; charset=utf-8",
             "strict-transport-security": "max-age=31536000",
             "connection": "close"
           },
           "httpVersion": "1.1"
         }
```

**Comment by viceice** [comment]
Remove auth type from your host rules

**Reply by viceice** [reply]
See https://github.com/renovatebot/renovate/discussions/10188 for quay auth issues.

**Reply by amimas** [reply]
Thank you @viceice for the suggestion.

I removed the `authType` and `token`. Used `username` and `password` for authenticating with docker registries. This time I didn't get those authentication error. My bot config now looks like this:

```
{
  platform: 'gitlab',
  endpoint: 'https://***/api/v4/',
  token: '***',
  autodiscover: true,
  binarySource: 'docker',
  gitAuthor: '',
  force: { automerge: false, enabled: true },
  logFile: 'renovate.log',
  logFileLevel: 'debug',
  onboarding: false,
  optimizeForDisabled: true,
  persistRepoData: false,
  prCommitsPerRunLimit: 0,
  printConfig: true,
  requireConfig: false,
  secrets: {
    DOCKERHUB_AUTH_USERNAME: '***',
    DOCKERHUB_AUTH_PASSWORD: '***',
    QUAY_AUTH_USERNAME: '***',
    QUAY_AUTH_PASSWORD: '***'
  },
  extends: [ 'config:base' ],
  assigneesFromCodeOwners: true,
  automerge: false,
  commitBodyTable: true,
  enabled: true,
  hostRules: [
    {
      matchHost: 'index.docker.io',
      username: '{{ secrets.DOCKERHUB_AUTH_USERNAME }}',
      password: '{{ secrets.DOCKERHUB_AUTH_PASSWORD }}'
    },
    {
      matchHost: 'quay.io',
      username: '{{ secrets.QUAY_AUTH_USERNAME }}',
      password: '{{ secrets.QUAY_AUTH_PASSWORD }}'
    }
  ],
  prFooter: 'This MR has been generated by [Renovate Bot](https://***/renovatebot/)'
}
```

So far I'm running this in dry-mode with only a handful of projects. Once I include more projects to be scanned, I will have more data to confirm if everything is okay. But the dry-run seems okay now.

I only had Personal Access Token from docker hub in my initial setup when I was using `token` and `authType` configs. I don't think these will work with Personal Access Token. Please correct me if I'm wrong but I seem to have avoided the authentication errors after switching to username and password.

**Reply by amimas** [reply]
Looks like I can only select 1 thread as answer in this discussion.

**Reply by viceice** [reply]
I think your quay hostrule won't work, as the quay `api/v1` only supports token auth. at least their docs say you need a token.

**Reply by amimas** [reply]
You are probably right. I will need to test it out a bit more.

> I don't think these will work with Personal Access Token.

☝️ - What I said earlier is not entirely accurate. For public docker hub registry, I am using `username` and `password` fields in renovate config. But the value of `password` is actually the Personal Access Token of my account from docker hub. With this change, I did not run into the authentication errors I posted before.

How do I validate that authentication was attempted and successful? I'm not seeing any errors for  either docker hub or quay registry in the console output of the bot's run. Would there be more info in the log file?

**Comment by rarkins** [comment, marked as answer]
Fix for secrets validation: #10206

**Reply by amimas** [reply]
Thank you @rarkins . The config validator now works after upgrading to latest version.
